### PR TITLE
DOC: minimize_ipopt: add documentation

### PR DESCRIPTION
**Pull request by mdhaber**
Thanks for maintaining cyipopt! 

I'm a SciPy maintainer, and we've discussed the possibility of adding `method='ipopt'` to `scipy.optimize.minimize` (with cyipopt as an optional dependency due to the license incompatibility). For now, though, we thought we'd try to help improve the compatibility between `cyipopt.minimize_ipopt` and `scipy.optimize.minimize` interfaces.

To start, this PR begins to add formal documentation to `cyipopt.minimize_ipopt`.

Please let me know if you're interested in continued contributions. In this PR, for example, I could also try to add:
- More complete documentation of `constraints`
- References (e.g. Ipopt documentation)
- Examples

As follow-up work, ideas include:
- Support for the `method` parameter (i.e. allowing use of all `scipy.optimize.minimize` methods)
- Support for SciPy's new-style constraints ([`Bounds`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.Bounds.html), [`NonlinearConstraint`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.NonlinearConstraint.html#scipy.optimize.NonlinearConstraint))
- Trying to improve numerical derivatives. (as-is, I'm not having very good luck unless analytical derivatives are defined.)

Please let me know if you'd like me to file issues for problems I've noticed. For instance, I'm on Windows and after creating a fresh environment, `conda install -c conda-forge cyipopt jax` fails due to an environment conflict.

### CODE REVIEW

File: cyipopt/scipy_interface.py [hunk 1]
```diff
@@ -324,11 +324,77 @@ def minimize_ipopt(fun,
...
+        ndarray, shape(n, )``. If ``None``, SciPy's ``approx_fprime`` is used.
+    hess : callable, optional
+        If ``None``, the Hessian is computed using IPOPT's numerical methods.
+        Explicitly defined Hessians are not yet supported via this interface.
+    hessp : callable, optional
+        If ``None``, the Hessian is computed using IPOPT's numerical methods.
+        Explicitly defined Hessians are not yet supported via this interface.
+    bounds :  sequence, shape(n, ), optional
+        Sequence of ``(min, max)`` pairs for each element in `x`. Use ``None``
+        to specify no bound.
```
Comment (mdhaber) [inline cyipopt/scipy_interface.py:379]: I could add support for `Bounds` (separate PR).

File: cyipopt/scipy_interface.py [hunk 2]
```diff
@@ -324,11 +324,77 @@ def minimize_ipopt(fun,
...
+        See `scipy.optimize.minimize` for more information. Note that the
+        Jacobian of each constraint corresponds to the ``'jac'`` key and must
+        be a callable function with signature
+        ``jac(x) -> {ndarray, coo_array}``. If the constraint's
+        value of ``'jac'`` is ``True``, the constraint function ``fun`` must
+        return a tuple ``(con_val, con_jac)`` consisting of the evaluated
+        constraint ``con_val`` and the evaluated Jacobian ``con_jac``.
+    tol : float, optional (default=1e-8)
+        The desired relative convergence tolerance, passed as an option to
+        Ipopt. See [https://coin-or.github.io/Ipopt/OPTIONS.html] for details.
```
Comment (mdhaber) [inline cyipopt/scipy_interface.py:390]: If desired, I could add Ipopt documentation as a reference and make this
```suggestion
        Ipopt. See [1]_ for details.
```

File: cyipopt/scipy_interface.py [hunk 3]
```diff
@@ -324,11 +324,77 @@ def minimize_ipopt(fun,
...
+    options : dict, optional
+        A dictionary of solver options. The options ``disp`` and ``maxiter``
+        are automatically mapped to their Ipopt equivalents ``print_level``
+        and ``max_iter``. All other options are passed directly to Ipopt. See
+        [https://coin-or.github.io/Ipopt/OPTIONS.html] for details.
+    callback : callable, optional
+        This parameter is ignored.
```
Comment (mdhaber) [inline cyipopt/scipy_interface.py:397]: I could raise a warning if this is passed noting that it will be ignored.

**Review by mdhaber — COMMENTED** [review]
Some self-review.

**Comment by moorepants** [comment]
This all sounds good. I'm happy to accept PRs improving the scipy style interface.

> For instance, I'm on Windows and after creating a fresh environment, conda install -c conda-forge cyipopt jax fails due to an environment conflict.

Please open an issue on the cyipopt feedstock, We have pretty loose pins https://github.com/conda-forge/cyipopt-feedstock/blob/main/recipe/meta.yaml#L22, so I'm not sure what conflicts with jax.

File: cyipopt/scipy_interface.py [hunk 4]
```diff
@@ -324,11 +324,77 @@ def minimize_ipopt(fun,
...
+        Initial guess. Array of real elements of shape (n,),
+        where ``n`` is the number of independent variables.
+    args : tuple, optional
+        Extra arguments passed to the objective function and its
+        derivatives (``fun``, ``jac``, and ``hess``).
+    kwargs : dictionary, optional
+        Extra keyword arguments passed to the objective function and its
+        derivatives (``fun``, ``jac``, ``hess``).
+    method : str, optional
+        This parameter is ignored. `minimize_ipopt` always uses Ipopt; use
```
[the remark this answers is not in the document]
Comment (moorepants) [reply cyipopt/scipy_interface.py:366]: I don't see anything wrong with supporting this.

File: cyipopt/scipy_interface.py [hunk 5]
```diff
@@ -324,11 +324,77 @@ def minimize_ipopt(fun,
...
+        `scipy.optimize.minimize` directly for other methods.
+    jac : callable, optional
+        The Jacobian of the objective function: ``jac(x, *args, **kwargs) ->
+        ndarray, shape(n, )``. If ``None``, SciPy's ``approx_fprime`` is used.
+    hess : callable, optional
+        If ``None``, the Hessian is computed using IPOPT's numerical methods.
+        Explicitly defined Hessians are not yet supported via this interface.
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply cyipopt/scipy_interface.py]: Please check me, but it looks like it was added in gh-121. In the meantime, I'll update this with the required signature (which does not allow *args, **kwargs, apparently).

**Review by mdhaber — COMMENTED** [review]
Thanks @moorepants! I made the documentation changes I mentioned, so I think this is ready for review. I can't build the documentation locally, though - is there a way of checking that it renders correctly?

**Comment by moorepants** [comment]
> I can't build the documentation locally, though - is there a way of checking that it renders correctly?

You'd have to build it locally or merge this PR and look at the latest build on readthedocs.

**Comment by moorepants** [comment]
> Please check me, but it looks like it was added in https://github.com/mechmotum/cyipopt/pull/121. In the meantime, I'll update this with the required signature (which does not allow *args, **kwargs, apparently).

Looks like it may take `*args` but we'd have to test it out. I'm not seeing an example. https://github.com/mechmotum/cyipopt/blob/master/cyipopt/scipy_interface.py#L207

**Comment by moorepants** [comment]
The args and kwargs seem to work if provided in the constraints dictionary:

https://github.com/mechmotum/cyipopt/pull/197

**Comment by moorepants** [comment]
I take that back, there is a failure.

**Comment by moorepants** [comment]
I think I fixed the args/kwargs things in #197, but I won't do anything with the docs there, just include the fix.

**Comment by mdhaber** [comment]
Do I understand that you would like to merge gh-197 first, then update the documentation of `hess` here (rather than merge this, then add the functionality and update the documentation of `hess` there)?

**Comment by tupui** [comment]
@mdhaber You mean the refs? For that you need numpydoc

**Comment by tupui** [comment]
Ah ok I see! I will have a look, did something similar for SALib. You should not need to add external etc.

**Comment by tupui** [comment]
Oh I think it's just the intershpinx link for SciPy. In SALib I have something simpler for the url.

File: docs/source/conf.py [hunk 6]
```diff
@@ -255,3 +256,11 @@
 
 # How to display URL addresses: 'footnote', 'no', or 'inline'.
 #texinfo_show_urls = 'footnote'
+
+# -----------------------------------------------------------------------------
+# Intersphinx configuration
+# -----------------------------------------------------------------------------
+intersphinx_mapping = {
+    'python': ('https://docs.python.org/dev', None),
+    'scipy': ('https://docs.scipy.org/doc/scipy/reference', None),
```
Comment (tupui) [inline docs/source/conf.py:265]:
```suggestion
    'scipy': ('https://docs.scipy.org/doc/scipy', None),
```

**Comment by moorepants** [comment]
This looks good to me. I'll merge and we can tidy anything else in other PRs.

**Comment by moorepants** [comment]
The references to scipy don't seem to work in the live version: https://cyipopt.readthedocs.io/en/latest/reference.html

**Comment by tupui** [comment]
Yes this is what Matt was saying above. You could try my suggestion above locally and see if when you build the doc it's solving the issue.

**Comment by moorepants** [comment]
Fixed here 3a69165

**Comment by moorepants** [comment]
Sorry, I didn't recognize that there were two people discussing this and that Matt hadn't come back to it. I corrected it. Sorry for the premature merge.

**Comment by tupui** [comment]
No worries 😃

**Pull request merged**